servoshell: show a status bar tooltip when hovering over links

MunishMummadi opened the pull request:
### Description
Adds functionality to show a status bar tooltip when hovering over links in the minibrowser. This PR watches for `EmbedderMsg::Status` events and keeps track of the current status value in the `Minibrowser`. When the page has status text, an egui tooltip is displayed at the bottom left corner.

### Changes Made
- Added a new field `status_text: Option<String>` to the `Minibrowser` struct to store the current status text.
- Updated `Minibrowser` methods:
  - `queue_embedder_events_for_minibrowser_events` now listens for `EmbedderMsg::Status` events and updates the `status_text` accordingly.
  - Updated `update` method to display a tooltip when `status_text` is present.
- Added a tooltip to the bottom left corner using `ui.tooltip_at_rect` in the `TopBottomPanel` show routine.

### Checklist
- [x] Created `status_text` field in `Minibrowser` struct.
- [x] Updated `queue_embedder_events_for_minibrowser_events` to handle `EmbedderMsg::Status`.
- [x] Added logic in `update` to show a tooltip when `status_text` is present.
- [x] Added tooltip display in the `TopBottomPanel` show routine.

### Related Issues
Closes #31690

MunishMummadi committed:
Fix typo: changed seperator to separator

MunishMummadi committed:
Merge branch 'servo:main' into main

MunishMummadi committed:
status-bar-tooltip-31690

MunishMummadi commented:
hey @delan . Can you review my PR. Guidance is appreciated.

eerii commented on ports/servoshell/minibrowser.rs:
```diff
@@ -1,7 +1,3 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
```
All the comments seem to be deleted. This is probably because your ide formatted the file on save, so maybe try to take a look into the settings to preserve comments, and then you can add them back. There's also very helpful documentation [here](https://github.com/servo/servo/blob/main/docs/HACKING_QUICKSTART.md#editor-support) to set up your editor of choice.

MunishMummadi replied:
thanks for the input. I resolved that.

eerii commented on ports/servoshell/minibrowser.rs:
```diff
@@ -189,64 +167,76 @@ impl Minibrowser {
                     let size = ui.available_size();
                     let rect = egui::Rect::from_min_size(min, size);
                     ui.allocate_space(size);
-
```
Try to run `./mach fmt` just before you make the commit, since it is failing the automatic checks because of formatting. You can use `./mach test-tidy` to verify that everything is fine before pushing.

MunishMummadi replied:
thanks for the input. 
```
munish@Zoro:~/servo$ ./mach fmt
munish@Zoro:~/servo$ ./mach test-tidy
 ➤  Checking config file (./servo-tidy.toml)...
 ➤  Checking directories for correct file extensions...
 ➤  Checking cargo lock (/home/munish/servo/Cargo.lock)...
 ➤  Checking files for tidiness
 ➤  Skipping WPT lint checks, because no relevant files changed.
 ➤  Checking formatting of rust files...
 ➤  Checking formatting of toml files...

 ✅ test-tidy reported no errors.
munish@Zoro:~/servo$
```

eerii reviewed (commented):
It also seems that the first commit was from #31685. If it's possible it would be great if you could rebase the changes without that commit so that there are no conflicts.

MunishMummadi committed:
updated the comments

MunishMummadi commented:
sir @eerii . Am I good to go. I tried to rebase but failed. It won't repeat again. Thank you for the guidance.

delan commented:
[The changes in this patch](https://github.com/servo/servo/pull/31747/files/9888ee7eb2e97e0204f8bf8a5901614eeed76be4) don’t match the description. For example, there is a new hovered_link field, but no status_text, and there are no changes to how we handle EmbedderMsg::Status. Have you pushed your latest changes?

> sir @<!---->eerii . Am I good to go. I tried to rebase but failed. It won't repeat again. Thank you for the guidance.

In the Servo project, please avoid using honorifics like “sir” and “madam”. I understand this can be a cultural difference, but not everyone is comfortable being addressed that way, and your message would still read as polite without it. You can refer to people by name, @mention, or their pronouns, all of these are ok.

MunishMummadi commented:
Hey @delan,

I apologize for any confusion caused. I've implemented the changes as discussed. However, I've encountered some issues and could use your assistance in the following areas: 

1. **Import Issue with `EmbedderMsg`:**
   In the `minibrowser.rs` file, I've attempted to import `EmbedderMsg` using the statement `use servo::compositing::windowing::EmbedderMsg;`. Could you please confirm if this is the correct import statement?

2.**Method Argument Mismatch:**
Additionally, I'm encountering issues with method calls in the minibrowser.rs file. The compiler reports that methods like output() and fonts() are being called without the expected arguments. It seems like I'm missing some arguments in these method calls. Could you provide guidance on how to correctly supply the required arguments to these methods?

The complete code I came up with is 
```
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

 use std::cell::{Cell, RefCell};
 use std::num::NonZeroU32;
 use std::sync::Arc;
 use std::time::Instant;
 
 use egui::{CentralPanel, Frame, InnerResponse, Key, Modifiers, PaintCallback, TopBottomPanel};
 use egui_glow::CallbackFn;
 use egui_winit::EventResponse;
 use euclid::{Length, Point2D, Scale};
 use gleam::gl;
 use glow::NativeFramebuffer;
 use log::{trace, warn};
 use servo::compositing::windowing::EmbedderEvent;
 use servo::msg::constellation_msg::TraversalDirection;
 use servo::rendering_context::RenderingContext;
 use servo::servo_geometry::DeviceIndependentPixel;
 use servo::servo_url::ServoUrl;
 use servo::compositing::windowing::EmbedderMsg;

 use crate::egui_glue::EguiGlow;
 use crate::events_loop::EventsLoop;
 use crate::geometry::winit_position_to_euclid_point;
 use crate::parser::location_bar_input_to_url;
 use crate::webview::WebViewManager;
 use crate::window_trait::WindowPortsMethods;

 pub struct Minibrowser {
     pub context: EguiGlow,
     pub event_queue: RefCell<Vec<MinibrowserEvent>>,
     pub toolbar_height: Length<f32, DeviceIndependentPixel>,
 
     /// The framebuffer object name for the widget surface we should draw to, or None if our widget
     /// surface does not use a framebuffer object.
     widget_surface_fbo: Option<NativeFramebuffer>,
 
     last_update: Instant,
     last_mouse_position: Option<Point2D<f32, DeviceIndependentPixel>>,
     location: RefCell<String>,
 
     /// Whether the location has been edited by the user without clicking Go.
     location_dirty: Cell<bool>,
     status_text: Option<String>, // Add this field
 }
 
 pub enum MinibrowserEvent {
     /// Go button clicked.
     Go,
     Back,
     Forward,
     Status(String), // Add Status variant to hold status text
 }

 impl Minibrowser {
     pub fn new(
         rendering_context: &RenderingContext,
         events_loop: &EventsLoop,
         window: &dyn WindowPortsMethods,
         initial_url: ServoUrl,
     ) -> Self {
         let gl = unsafe {
             glow::Context::from_loader_function(|s| rendering_context.get_proc_address(s))
         };
 
         // Adapted from https://github.com/emilk/egui/blob/9478e50d012c5138551c38cbee16b07bc1fcf283/crates/egui_glow/examples/pure_glow.rs
         let context = EguiGlow::new(events_loop.as_winit(), Arc::new(gl), None);
         context
             .egui_ctx
             .set_pixels_per_point(window.hidpi_factor().get());
 
         let widget_surface_fbo = match rendering_context.context_surface_info() {
             Ok(Some(info)) => NonZeroU32::new(info.framebuffer_object).map(NativeFramebuffer),
             Ok(None) => panic!("Failed to get widget surface info from surfman!"),
             Err(error) => panic!("Failed to get widget surface info from surfman! {error:?}"),
         };

         Self {
             context,
             event_queue: RefCell::new(vec![]),
             toolbar_height: Default::default(),
             widget_surface_fbo,
             last_update: Instant::now(),
             last_mouse_position: None,
             location: RefCell::new(initial_url.to_string()),
             location_dirty: false.into(),
             status_text: None, //Initialize status_text field to none
         }
     }

     /// Preprocess the given [winit::event::WindowEvent], returning unconsumed for mouse events in
     /// the Servo browser rect. This is needed because the CentralPanel we create for our webview
     /// would otherwise make egui report events in that area as consumed.
     pub fn on_event(&mut self, event: &winit::event::WindowEvent<'_>) -> EventResponse {
         let mut result = self.context.on_event(event);
         result.consumed &= match event {
             winit::event::WindowEvent::CursorMoved { position, .. } => {
                 let scale = Scale::<_, DeviceIndependentPixel, _>::new(
                     self.context.egui_ctx.pixels_per_point(),
                 );
                 self.last_mouse_position =
                     Some(winit_position_to_euclid_point(*position).to_f32() / scale);
                 self.last_mouse_position
                     .map_or(false, |p| self.is_in_browser_rect(p))
             },
             winit::event::WindowEvent::MouseWheel { .. } |
             winit::event::WindowEvent::MouseInput { .. } => self
                 .last_mouse_position
                 .map_or(false, |p| self.is_in_browser_rect(p)),
             _ => true,
         };
         result
     }
 
     /// Return true iff the given position is in the Servo browser rect.
     fn is_in_browser_rect(&self, position: Point2D<f32, DeviceIndependentPixel>) -> bool {
         position.y < self.toolbar_height.get()
     }

     pub fn handle_embedder_msg(&mut self, msg: EmbedderMsg) {
        match msg {
            EmbedderMsg::Status(status) => {
                self.event_queue.borrow_mut().push(MinibrowserEvent::Status(status));
            }
            // Handle other EmbedderMsg variants if needed
            _ => {}
        }
    }

     /// Update the minibrowser, but don’t paint.
     /// If `servo_framebuffer_id` is given, set up a paint callback to blit its contents to our
     /// CentralPanel when [`Minibrowser::paint`] is called.
     pub fn update(
         &mut self,
         window: &winit::window::Window,
         servo_framebuffer_id: Option<gl::GLuint>,
         reason: &'static str,
     ) {
         let now = Instant::now();
         trace!(
             "{:?} since last update ({})",
             now - self.last_update,
             reason
         );
         let Self {
             context,
             event_queue,
             toolbar_height,
             widget_surface_fbo,
             last_update,
             location,
             location_dirty,
             ..
         } = self;
         let widget_fbo = *widget_surface_fbo;
         let _duration = context.run(window, |ctx| {
             let InnerResponse { inner: height, .. } =
                 TopBottomPanel::top("toolbar").show(ctx, |ui| {
                     ui.allocate_ui_with_layout(
                         ui.available_size(),
                         egui::Layout::left_to_right(egui::Align::Center),
                         |ui| {
                             if ui.button("back").clicked() {
                                 event_queue.borrow_mut().push(MinibrowserEvent::Back);
                             }
                             if ui.button("forward").clicked() {
                                 event_queue.borrow_mut().push(MinibrowserEvent::Forward);
                             }
                             ui.allocate_ui_with_layout(
                                 ui.available_size(),
                                 egui::Layout::right_to_left(egui::Align::Center),
                                 |ui| {
                                     if ui.button("go").clicked() {
                                         event_queue.borrow_mut().push(MinibrowserEvent::Go);
                                         location_dirty.set(false);
                                     }
 
                                     let location_field = ui.add_sized(
                                         ui.available_size(),
                                         egui::TextEdit::singleline(&mut *location.borrow_mut()),
                                     );
 
                                     if location_field.changed() {
                                         location_dirty.set(true);
                                     }
                                     if ui.input(|i| {
                                         i.clone().consume_key(Modifiers::COMMAND, Key::L)
                                     }) {
                                         location_field.request_focus();
                                     }
                                     if location_field.lost_focus() &&
                                         ui.input(|i| i.clone().key_pressed(Key::Enter))
                                     {
                                         event_queue.borrow_mut().push(MinibrowserEvent::Go);
                                         location_dirty.set(false);
                                     }
                                 },
                             );
                         },
                     );
                     ui.cursor().min.y
                 });
             *toolbar_height = Length::new(height);
 
             CentralPanel::default()
                 .frame(Frame::none())
                 .show(ctx, |ui| {
                     let min = ui.cursor().min;
                     let size = ui.available_size();
                     let rect = egui::Rect::from_min_size(min, size);
                     ui.allocate_space(size);
 
                     let Some(servo_fbo) = servo_framebuffer_id else {
                         return;
                     };
                     ui.painter().add(PaintCallback {
                         rect,
                         callback: Arc::new(CallbackFn::new(move |info, painter| {
                             use glow::HasContext as _;
                             let clip = info.viewport_in_pixels();
                             let x = clip.left_px as gl::GLint;
                             let y = clip.from_bottom_px as gl::GLint;
                             let width = clip.width_px as gl::GLsizei;
                             let height = clip.height_px as gl::GLsizei;
                             unsafe {
                                 painter.gl().clear_color(0.0, 0.0, 0.0, 0.0);
                                 painter.gl().scissor(x, y, width, height);
                                 painter.gl().enable(gl::SCISSOR_TEST);
                                 painter.gl().clear(gl::COLOR_BUFFER_BIT);
                                 painter.gl().disable(gl::SCISSOR_TEST);
 
                                 let servo_fbo = NonZeroU32::new(servo_fbo).map(NativeFramebuffer);
                                 painter
                                     .gl()
                                     .bind_framebuffer(gl::READ_FRAMEBUFFER, servo_fbo);
                                 painter
                                     .gl()
                                     .bind_framebuffer(gl::DRAW_FRAMEBUFFER, widget_fbo);
                                 painter.gl().blit_framebuffer(
                                     x,
                                     y,
                                     x + width,
                                     y + height,
                                     x,
                                     y,
                                     x + width,
                                     y + height,
                                     gl::COLOR_BUFFER_BIT,
                                     gl::NEAREST,
                                 );
                                 painter.gl().bind_framebuffer(gl::FRAMEBUFFER, widget_fbo);
                             }
                         })),
                     });
                 });
 
             *last_update = now;
         });
         // Update status text and trigger repaint if necessary
        self.update_status_text_and_repaint();
     }
 
     /// Paint the minibrowser, as of the last update.
     pub fn paint(&mut self, window: &winit::window::Window) {
         unsafe {
             use glow::HasContext as _;
             self.context
                 .painter
                 .gl()
                 .bind_framebuffer(gl::FRAMEBUFFER, self.widget_surface_fbo);
         }
         self.context.paint(window);
     }
 
     /// Takes any outstanding events from the [Minibrowser], converting them to [EmbedderEvent] and
     /// routing those to the App event queue.
     pub fn queue_embedder_events_for_minibrowser_events(
         &self,
         browser: &WebViewManager<dyn WindowPortsMethods>,
         app_event_queue: &mut Vec<EmbedderEvent>,
     ) {
         for event in self.event_queue.borrow_mut().drain(..) {
             match event {
                 MinibrowserEvent::Go => {
                     let browser_id = browser.webview_id().unwrap();
                     let location = self.location.borrow();
                     if let Some(url) = location_bar_input_to_url(&location.clone()) {
                         app_event_queue.push(EmbedderEvent::LoadUrl(browser_id, url));
                     } else {
                         warn!("failed to parse location");
                         break;
                     }
                 },
                 MinibrowserEvent::Back => {
                     let browser_id = browser.webview_id().unwrap();
                     app_event_queue.push(EmbedderEvent::Navigation(
                         browser_id,
                         TraversalDirection::Back(1),
                     ));
                 },
                 MinibrowserEvent::Forward => {
                     let browser_id = browser.webview_id().unwrap();
                     app_event_queue.push(EmbedderEvent::Navigation(
                         browser_id,
                         TraversalDirection::Forward(1),
                     ));
                 },
             }
         }
     }

     /// Updates the location field from the given [BrowserManager], unless the user has started
     /// editing it without clicking Go, returning true iff it has changed (needing an egui update).
     pub fn update_location_in_toolbar(
         &mut self,
         browser: &mut WebViewManager<dyn WindowPortsMethods>,
     ) -> bool {
         // User edited without clicking Go?
         if self.location_dirty.get() {
             return false;
         }

         match browser.current_url_string() {
             Some(location) if location != self.location.get_mut() => {
                 self.location = RefCell::new(location.to_owned());
                 true
             },
             _ => false,
         }
     }
     /// Update status text and trigger repaint if necessary
    fn update_status_text_and_repaint(&mut self) {
        if self.status_text.is_some() {
            // Trigger repaint if status text is present
            self.context.egui_ctx.request_repaint();
        }
    }
    // Method to update status text
    pub fn update_status_text(&mut self, status: Option<String>) {
        self.status_text = Some(status);
        self.context.egui_ctx.request_repaint();
    }

    /// Show tooltip if status text is present.
    pub fn show_status_tooltip(&self, ui: &mut egui::Ui) {
        if let Some(status_text) = &self.status_text {
            let max_y = ui.cursor().max.y;
            ui.ctx().output().tooltip_text(
                egui::Id::new("status_tooltip"),
                egui::Rect::from_min_size(
                    egui::Pos2::new(0.0, max_y),
                    egui::Vec2::new(ui.available_width(), ui.fonts().row_height()),
                ),
                status_text.clone(),
            );
        }
    }
 }
```

Looking forward to your feedback and guidance.

gterzian commented:
> The compiler reports that 

Could you please post the error?

MunishMummadi commented:
> > The compiler reports that
> 
> Could you please post the error?

follow is error I got after running `./mach build `

```
munish@Zoro:~/servo$ ./mach build
No build type specified so assuming `--dev`.
   Compiling servoshell v0.0.1 (/home/munish/servo/ports/servoshell)
error[E0432]: unresolved import `servo::compositing::windowing::EmbedderMsg`
  --> ports/servoshell/minibrowser.rs:22:6
   |
22 |  use servo::compositing::windowing::EmbedderMsg;
   |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `EmbedderMsg` in `windowing`

error[E0308]: mismatched types
   --> ports/servoshell/minibrowser.rs:342:33
    |
342 |         self.status_text = Some(status);
    |                            ---- ^^^^^^ expected `String`, found `Option<String>`
    |                            |
    |                            arguments to this enum variant are incorrect
    |
    = note: expected struct `std::string::String`
                 found enum `std::option::Option<std::string::String>`
help: the type constructed contains `std::option::Option<std::string::String>` due to the type of the argument passed
   --> ports/servoshell/minibrowser.rs:342:28
    |
342 |         self.status_text = Some(status);
    |                            ^^^^^------^
    |                                 |
    |                                 this argument influences the type of `Some`
note: tuple variant defined here
   --> /rustc/a28077b28a02b92985b3a3faecf92813155f1ea1/library/core/src/option.rs:571:5

error[E0061]: this method takes 1 argument but 0 arguments were supplied
   --> ports/servoshell/minibrowser.rs:350:22
    |
350 |             ui.ctx().output().tooltip_text(
    |                      ^^^^^^-- an argument is missing
    |
note: method defined here
   --> /home/munish/.cargo/registry/src/index.crates.io-6f17d22bba15001f/egui-0.22.0/src/context.rs:468:12
    |
468 |     pub fn output<R>(&self, reader: impl FnOnce(&PlatformOutput...
    |            ^^^^^^
help: provide the argument
    |
350 |             ui.ctx().output(/* reader */).tooltip_text(
    |                            ~~~~~~~~~~~~~~

error[E0061]: this method takes 1 argument but 0 arguments were supplied
   --> ports/servoshell/minibrowser.rs:354:62
    |
354 | ...le_width(), ui.fonts().row_height()),
    |                   ^^^^^-- an argument is missing
    |
note: method defined here
   --> /home/munish/.cargo/registry/src/index.crates.io-6f17d22bba15001f/egui-0.22.0/src/ui.rs:417:12
    |
417 |     pub fn fonts<R>(&self, reader: impl FnOnce(&Fonts) -> R) ->...
    |            ^^^^^
help: provide the argument
    |
354 |                     egui::Vec2::new(ui.available_width(), ui.fonts(/* reader */).row_height()),
    |
~~~~~~~~~~~~~~

Some errors have detailed explanations: E0061, E0308, E0432.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `servoshell` (lib) due to 4 previous errors
      Timing report saved to /home/munish/servo/target/cargo-timings/cargo-timing-20240320T010049Z.html
Failed in 0:00:07
[Warning] Could not generate notification: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.Notifications was not provided bymunish@Zoro:~/servo$
```

gterzian commented:
> use servo::compositing::windowing::EmbedderMsg;

`use embedder_traits::EmbedderMsg;`

> self.status_text = Some(status);

Did you change `status_text` into an option? If that is intended, I guess you have to update the type on the definition of `self`. 

For the others, the errors are from the `ui` library, did you inadvertently update the version or something? Unless you changed the existing code I don't see why it would start failing.

MunishMummadi commented:
Hey @gterzian  I am able to get rid of the first error. I will work on the remaining things too. I need a clarification about my approach. whether my approach is in the correct direction or not. I don't expect you to look at the complete code. Thanks for the help though.

gterzian commented:
With regard to the overall approach, I will refer to @delan comment above: 

> [The changes in this patch](https://github.com/servo/servo/pull/31747/files/9888ee7eb2e97e0204f8bf8a5901614eeed76be4) don’t match the description. For example, there is a new hovered_link field, but no status_text, and there are no changes to how we handle EmbedderMsg::Status. Have you pushed your latest changes?

Have those comments been addressed?

MunishMummadi commented:
> > I am able to get rid of the first error. I will work on the remaining things too.
> 
> Good!
> 
> > whether my approach is in the correct direction or not.
> 
> With regard to the overall approach, I will refer to @delan comment above:
> 
> > [The changes in this patch](https://github.com/servo/servo/pull/31747/files/9888ee7eb2e97e0204f8bf8a5901614eeed76be4) don’t match the description. For example, there is a new hovered_link field, but no status_text, and there are no changes to how we handle EmbedderMsg::Status. Have you pushed your latest changes?
> 
> Have those comments been addressed?

No @gterzian I haven't pushed the latest changes. Because I am struck at what arguments to pass. Do you want me to push them.

MunishMummadi committed:
added-embeddermsg-status-tooltip

gterzian commented on ports/servoshell/minibrowser.rs:
```diff
@@ -314,4 +331,31 @@ impl Minibrowser {
+    pub fn show_status_tooltip(&self, ui: &mut egui::Ui) {
+        if let Some(status_text) = &self.status_text {
+            let max_y = ui.cursor().max.y;
+            ui.ctx().output().tooltip_text(
```
`output` seems to take a closure: https://docs.rs/egui/latest/egui/struct.Context.html#method.output. I think you can try to add a tooltip using the [`PlatformOutput`](https://docs.rs/egui/latest/egui/output/struct.PlatformOutput.html) passed as the argument to the closure.

gterzian replied:
Similar for fonts: https://docs.rs/egui/latest/egui/struct.Context.html#method.fonts

gterzian reviewed (changes requested):
The two methods take closures as arguments.

MunishMummadi committed:
Got-rif-of-errors

MunishMummadi commented:
> The two methods take closures as arguments.

Hey @gterzian . I missed a simple thing. I updated it . Can you have look at it.

sandeepB3 commented on ports/servoshell/minibrowser.rs:
```diff
@@ -241,6 +251,19 @@ impl Minibrowser {
                 });
+            // Show status tooltip if there is status text
+            if let Some(status_text) = status_text_clone.as_ref() {
+                let screen_rect = ctx.available_rect();
```
I think ctx is not defined within this scope. I would suggest moving this block of code to a section where ctx is defined, such as after line 156.

sandeepB3 commented:
@MunishMummadi, before you push your code, I would suggest you to run `./mach build -d` and `./mach test-tidy` to ensure that building servo after your changes does not report any errors, since none of your PR pass the workflow checks, I have suggested a comment in your code which can cause `./mach build -d` to report an error.

MunishMummadi commented:
> @MunishMummadi, before you push your code, I would suggest you to run `./mach build -d` and `./mach test-tidy` to ensure that building servo after your changes does not report any errors, since none of your PR pass the workflow checks, I have suggested a comment in your code which can cause `./mach build -d` to report an error.

Hey @sandeepB3 . I ran `./mach fmt` and `./mach test-tidy. It hasn't reported any errors. I will run `./mach build -d`. thnx for the suggestion.

sandeepB3 commented:
I would also suggest you to run `./mach run -d -- <any link of your choice>`, then test if the tooltip works when hovering over links, since even in my draft PR `./mach build -d` and `./mach test-tidy` does not report any error, but opening any link in the minibrowser and hovering over a link does not show the tooltip.

MunishMummadi commented:
I am closing this PR as it became too messy. If possible I will link my latest updated changes in a New PR
